added new example (3d quadcopter)

cjmcgreal opened the pull request:
Matt,

I am still considering this "work in progress", but progress is slowing and so I created the pull request in case you were available for input / review.

I still have not been able to get it to converge to a result, despite it having run >3000 iterations. (trapz with n = 16).

I did some preliminary testing to check dynamics function for expected behavior.  (included).

A couple of thoughts on potential issues:
With the 6 DOF model, including the velocities that gives 12 states.  I am not sure what should be the 'expected' number of iterations for that problem 'size'.
I've (somewhat naively) implemented nested for-loops; I am not sure if these are playing well with the solver.  If you think this is the issue, and you have suggestions for improvement, please let me know.

I've tried to strike a balance between expediency and flexibility, documentation and verbosity.  Kindly let me know any and all thoughts you might have: suggestions, recommendations, code style, variable naming conventions etc.

Thanks again,
Conrad

cjmcgreal commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -0,0 +1,18 @@
+function [R] = Euler2RotMat(theta_x, theta_y, theta_z)
```
eul2rotm() is a function in the Robotic System Toolbox and would introduce a dependency where there currently isn't one.

I am happy to update/correct the implementation as per your recommendations.

I wrote this function, a while back so there shouldn't be any licensing issues.
I've updated the reference/citing so it's hopefully more readable.

cjmcgreal commented on demo/quadRotor3d/utilities/computePropOpPoint.m:
```diff
@@ -0,0 +1,27 @@
+function [thrust, torque] = computePropOpPoint(RPM, rho, d_prop, C_t, C_q)
```
Thanks!!

cjmcgreal commented:
Matt,

I'm pretty sure I've addressed all of your comments, but may have introduced new issues...

**vectorization**: still haven't completely gotten rid of the nested for loops, in particular the coordinate transform, i.e. call to Euler2RotMat:
One option is to use eul2rotm() and create the dependency on the Robotic System Toolbox.

Let me know what you think,

Conrad

MatthewPeterKelly commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -1,18 +1,39 @@
 
-R = Z*Y*X;
\ No newline at end of file
+    R(:,:,i) = Z*Y*X;
```
See above. You can likely speed this up by precomputing it with the symbolic toolbox. That will let you drop a lot of multiplies by 0 and 1.

MatthewPeterKelly commented on demo/quadRotor3d/utilities/plot_quadRotor3d.m:
```diff
@@ -0,0 +1,68 @@
+function plot_quadRotor3d(soln)
```
This looks great - thanks for adding!

cjmcgreal commented on demo/quadRotor3d/dynQuadrotor3D.m:
```diff
@@ -27,29 +27,30 @@
-    linAccel_world = linAccel_body' * BodyRotMat' ; % transform body frame linear acceleration into world frame 
-    angAccel_world = angAccel_body' * BodyRotMat' ; % transform body frame angular acceleration into world frame 
+for i=1:size(BodyRotMat,3)
+linAccel_world = linAccel_body(i,:) * BodyRotMat(:,:,i)' ; % transform body frame linear acceleration into world frame 
```
I will have to follow up on this.

I was under the impression that, in the body frame, one could neglect Coriolis terms for example, but now that you're drawing my attention to this, I'm realizing that was more from memory than from a specific reference so let me dig a little deeper here.

cjmcgreal replied:
I took a look at Etkin,B. Dynamics of Atmospheric Flight, and I think you're right that I need to include the bias acceleration terms.
In light of the complexity of this, I'm now planning to write a test that would compare the output of this dynamics function to the output of the Simulink Aerospace Toolbox 6DOF block.  This test itself would then of course have a dependency on both Simulink and the Aerospace Toolbox but obviously a user wouldn't be required to actually run the test.  
Let me know what you think.

cjmcgreal commented on demo/quadRotor3d/test/TEST_definePropulsionModel.m:
```diff
@@ -0,0 +1,4 @@
+
```
I went back and forth on this a bit.  Maybe you can help me clarify the best structure here...
In my work elsewhere, I've been modularizing plant models such that the aerodynamics, propulsion, inertial, and environmental parameters are all independent substructs in the main "plan model struct".  For this 3d quad example that structure is a little 'heavy duty' but for the next example I'm planning on (i.e. transition with aerodynamics) there are a number of advantages to staying organized and separating the parameters for aero modeling from the parameters for propulsion.
I guess what I'm getting at is that when the 'propulsion model' is the only substruct it looks a little odd but like I said, in the next example I'm planning to do a plant model top level struct that has following substructs: 
.aerodynamics
.propulsion
.inertial
.environment

But that still leaves the question as to what is most appropriate for this particular example....

MatthewPeterKelly replied:
That makes a lot of sense - good idea! I'm all for keeping things as modular as possible. I opened up your code in my editor and the overall architecture makes more sense now. I think that `defines` in the name confused me at the start. I had originally thought that it was somehow used to create a file or export parameters. I now see that it does three things:
1) create an example set of parameters 
2) uses those parameters to create a model
3) generates a set of diagnostic plots from that model

cjmcgreal commented on demo/quadRotor3d/test/TEST_computePropOpPoint.m:
```diff
@@ -0,0 +1,23 @@
+disp('Running: TEST_computePropOpPoint.m')
```
I added a little more on this, see file 'showPropulsionModel.m' (now called in TEST_definePropulsionModel.m)

On that note, I think I'd appreciate some curation here;  this is the type of stuff I look at all day long (i.e. propulsion system parameters etc) and so it's all very obvious to me what it is and what it's for and therefore it can be a little tricky for me to estimate the proper level of explanation and documentation... please feel free to tell me what you think is appropriate here.

cjmcgreal commented on demo/quadRotor3d/test/TEST_dynBodyFrame.m:
```diff
@@ -8,33 +8,38 @@
 %% Test 3 - only motor 1 (port side motor {at the moment}) 
 % should result in (pitch accel = 0, roll accel < 0, yaw accel < 0)
-disp('Test 3') 
+disp('Test 4 - port motor only ; ramp') 
```
awesome!

cjmcgreal commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -1,18 +1,39 @@
-function [R] = Euler2RotMat(theta_x, theta_y, theta_z)
+function [R] = Euler2RotMat(eul)
+% [R] = Euler2RotMat(theta_x, theta_y, theta_z)
```
yeah!

cjmcgreal commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -1,18 +1,39 @@
+R = zeros(3,3,size(eul,1)) ; 
 
-% See: http://nghiaho.com/?page_id=846 
+for i=1:size(eul,1) 
```
I took a stab at this (in an attempt to "be really fancy", in part for the learning experience of it) but am not at all sure if I'm on the right track!

So far I haven't made any changes to the dynamics function, but the Euler2RotMat function is returning a [9,N].  Is this even close to what you meant? :D

MatthewPeterKelly replied:
Great start! You're about half way there. I'll leave some suggestions inline. 

One big-picture thought about vectorization in Matlab: there are trade-offs. Usually it is faster, but for anything complicated it is good to verify it using the profiler. For example, if you need to copy or reshape big matrices then it is not a clear win. Another issue is that vectorized Matlab code is harder to read and understand, which is something that I struggled with when writing the core functions in this library - I wanted them to be as readable as possible, but also fast enough to be usable.

In this particular case, my guess is that if you carefully vectorize the code it will be faster but less understandable. Thus, I'm totally fine with you implementing it either way.

cjmcgreal commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -1,18 +1,39 @@
-    0 cos(theta_x) -sin(theta_x)
-    0 sin(theta_x) cos(theta_x)];
+    X = [1 0 0
+        0 cos(theta_x) -sin(theta_x)
```
easy fix

MatthewPeterKelly commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -16,24 +16,66 @@
 %
 % Written by Conrad McGreal @ 2019
-R = zeros(3,3,size(eul,1)) ; 
+R = zeros(9,size(eul,1)) ; 
```
Good.

MatthewPeterKelly commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -16,24 +16,66 @@
-R = zeros(3,3,size(eul,1)) ; 
+R = zeros(9,size(eul,1)) ; 
 
 for i=1:size(eul,1) 
```
This loop goes away in a vectorized version.

MatthewPeterKelly commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -16,24 +16,66 @@
     theta_y = eul(i,2) ; 
     theta_z = eul(i,3) ; 
 
+    cx = cos(theta_x) ; 
```
Exactly!  Once you drop the loop, `cx` will be a vector, rather than a scalar.

MatthewPeterKelly commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -16,24 +16,66 @@
+        R_square(9)] ;
+end
+
+function [C] = matrixMultiply(A,B)
```
This is the correct idea, but slightly off. Right now your calling this function inside the loop, but it needs to be outside the loop to get the speed-up. Thus, C should end up as [9, N], and `a,b,c,d, ...` should each be row vectors of length N.

MatthewPeterKelly commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -16,24 +16,66 @@
+o = B(8) ; 
+r = B(9) ; 
+
+C = [a*j+b*m+c*p, a*k+b*n+c*q, a*l+b*o+c*r
```
This looks good as the general solution, but you want the specific solution for your problem, which lets you drop many of these terms.

MatthewPeterKelly commented on demo/quadRotor3d/utilities/Euler2RotMat.m:
```diff
@@ -16,24 +16,66 @@
+        R_square(6)
+        R_square(7)
+        R_square(8)
+        R_square(9)] ;
```
Eventually, this line should read something like...
```matlab
R = vectorizedRotationMatrix(theta_x, theta_y, theta_z);
```
Then inside of `vectorizedRotationMatrix` should be a matlab verison of:    
![matrix multiply](https://user-images.githubusercontent.com/8137529/74393382-ee53d400-4dd7-11ea-829e-6876c07d33b7.png)

It looks like wikipedia lists out these rotation matrices in closed form as well:
https://en.wikipedia.org/wiki/Euler_angles

Where the 9 elements of the 3x3 matrix are rearranged to form a column vector (just like you did in your `matrixMultiply`) function.

MatthewPeterKelly commented on demo/quadRotor3d/MAIN.m:
```diff
@@ -6,7 +6,7 @@
 
 clc; clear;
-addpath ../../
+addpath ../../ ./utilities ./test
```
Thanks!

MatthewPeterKelly commented on demo/quadRotor3d/TESTRUNNER.m:
```diff
@@ -0,0 +1,14 @@
+% runs all tests in 'test' folder
+
+clc; clear;
+addpath ./test
```
Missing dependency on utilities. Need to do:
`addpath ./test ./utilities`

MatthewPeterKelly commented on demo/quadRotor3d/TESTRUNNER.m:
```diff
@@ -0,0 +1,14 @@
+% runs all tests in 'test' folder
```
This is a great idea - thanks for adding this script.

cjmcgreal commented:
Matt, 

I think I'm pretty happy with this PR at the moment.

Some comments:
'Hermite Simpson' - left this as 'optional' (i.e. commented it out).  As a new user running the example it would be nice to not have to wait so long.

plotting - Now evaluating using the built-in method. Added the optional argument. Added an optional argument to show the propulsion model while I was at it.

Looking forward to get started on the next example :)

Thanks!

MatthewPeterKelly commented on demo/quadRotor3d/utilities/definePropulsionModel.m:
```diff
@@ -55,4 +55,12 @@
+end
+
+% plot propulsion model, if desired
+if ~exist('plotflag','var')
```
This is a neat way to implement the default argument - I've not seen it before. The more common way to implement this is:
```matlab
if nargin < 2
    plotflag = false
```
Fine to leave code as-is; both implementations are fine.
